File: 백준/Platinum/4196. 도미노/도미노.py
# 도미노
import sys
input = sys.stdin.readline
sys.setrecursionlimit(100001)

def dfs(n):
    visited[n] = 1

    for i in graph[n]:
        if not visited[i]:
            dfs(i)

    order.append(n)

t = int(input())
for i in range(t):
    n, m = map(int,input().split())
    graph = [[] for _ in range(n+1)]
    graphR = [[] for _ in range(n+1)] # 역방향 그래프
    visited = [0] * (n+1)
    group = [0] * (n+1) # scc 그룹
    front = [0] * (n+1) # scc 앞에 있는 scc갯수
    order = []
    cnt = 1 # scc 갯수 카운팅
    result = 0 # 밀어야 하는 횟수
    
    for j in range(m):
        s, t = map(int,input().split())
        graph[s].append(t)
        graphR[t].append(s)

    for j in range(1, n+1):
        if not visited[j]:
            dfs(j)

    visited = [0] * (n+1)
    for j in range(n-1, -1, -1):
        if visited[order[j]]:
            continue
        
        stack = [order[j]]

        while stack:
            cur = stack.pop()
            if visited[cur]:
                continue
            else:
                visited[cur] = 1
                group[cur] = cnt

            for k in graphR[cur]:
                if visited[k]:
                    continue
                else:
                    stack.append(k)
        cnt += 1

    for j in range(1, n+1):
        for k in graph[j]:
            if group[j] == group[k]:
                continue
            else:
                front[group[k]] += 1

    for j in range(1, cnt):
        if front[j] == 0:
            result +=  1

    print(result)

    

File: 백준/Platinum/4196. 도미노/도미노.java
import java.io.*;
import java.util.*;

public class Main {
    static List<List<Integer>> graph, graphR;
    static boolean[] visited;
    static int[] group, front;
    static Stack<Integer> order;
    static int cnt;

    public static void dfs(int n) {
        visited[n] = true;
        for (int next : graph.get(n)) {
            if (!visited[next]) {
                dfs(next);
            }
        }
        order.push(n);
    }

    public static void dfsR(int n, int cnt) {
        visited[n] = true;
        group[n] = cnt;
        for (int next : graphR.get(n)) {
            if (!visited[next]) {
                dfsR(next, cnt);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());

            graph = new ArrayList<>();
            graphR = new ArrayList<>();
            for (int i = 0; i <= n; i++) {
                graph.add(new ArrayList<>());
                graphR.add(new ArrayList<>());
            }

            visited = new boolean[n + 1];
            group = new int[n + 1];
            front = new int[n + 1];
            order = new Stack<>();
            cnt = 1;

            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());
                int start = Integer.parseInt(st.nextToken());
                int to = Integer.parseInt(st.nextToken());
                graph.get(start).add(to);
                graphR.get(to).add(start);
            }

            for (int i = 1; i <= n; i++) {
                if (!visited[i]) {
                    dfs(i);
                }
            }

            Arrays.fill(visited, false);

            while (!order.isEmpty()) {
                int node = order.pop();
                if (!visited[node]) {
                    dfsR(node, cnt);
                    cnt++;
                }
            }

            for (int i = 1; i <= n; i++) {
                for (int next : graph.get(i)) {
                    if (group[i] != group[next]) {
                        front[group[next]]++;
                    }
                }
            }

            int result = 0;
            for (int i = 1; i < cnt; i++) {
                if (front[i] == 0) {
                    result++;
                }
            }

            sb.append(result).append("\n");
        }

        System.out.print(sb);
    }
}

